File: glovers_frontend/src/services/ProductsService.ts
import axios from 'axios'
import type IProduct from '@/interfaces/IProduct'
import type IReview from '@/interfaces/IReview'
import type IProductsService from '@/interfaces/IProductsService'

export default class ProductsService implements IProductsService {
  private baseUrl: string = 'https://localhost:44316/api/Products'

  private static defaultProductResponse: IProduct = {
    id: 0,
    title: '',
    price: 0,
    category: '',
    description: '',
    image: '',
    nutritionImage: ''
  }

  private static defaultReviewResponse: IReview = {
    id: 0,
    title: '',
    rating: 0,
    productId: -1,
    product: this.defaultProductResponse,
    description: ''
  }

  public async getAllProducts(): Promise<IProduct[]> {
    let response = await axios.get(`${this.baseUrl}/getAllProducts`)
    if (response.status == 200) {
      return response.data as IProduct[]
    } else {
      console.log(response, 'Failed to get all')
    }
    return [ProductsService.defaultProductResponse]
  }

  public async getProduct(id: number): Promise<IProduct> {
    let response = await axios.get(`${this.baseUrl}/getProduct/${id}`)
    if (response.status == 200) {
      return response.data as IProduct
    } else {
      console.log(response, `Failed to get product with id ${id}`)
    }
    return ProductsService.defaultProductResponse
  }

  public async getLimitedProducts(amount: number): Promise<IProduct[]> {
    let response = await axios.get(`${this.baseUrl}/getAllProducts/${amount}`)
    if (response.status == 200) {
      return response.data as IProduct[]
    } else {
      console.log(response, `Failed to get ${amount} products`)
    }
    return [ProductsService.defaultProductResponse]
  }

  public async getAllCategories(): Promise<string[]> {
    let defaultResp = ['']
    let response = await axios.get(`${this.baseUrl}/getAllCategories`)
    if (response.status == 200) {
      return response.data as string[]
    } else {
      console.log(response, `Failed to get products`)
    }
    return defaultResp
  }

  public async getAllProductsInCategory(category: string): Promise<IProduct[]> {
    let response = await axios.get(`${this.baseUrl}/getAllProducts/${category}`)
    if (response.status == 200) {
      return response.data as IProduct[]
    } else {
      console.log(response, `Failed to get all products in category: ${category}`)
    }
    return [ProductsService.defaultProductResponse]
  }

  public async getAllReviewForProduct(productId: number): Promise<IReview[]> {
    console.log()
    let response = await axios.get(`${this.baseUrl}/getReviews/${productId}`)
    if (response.status == 200) {
      return response.data as IReview[]
    } else {
      console.log(response, `Failed to get all reviews for product with id: ${productId}`)
    }
    return [ProductsService.defaultReviewResponse]
  }
}

File: glovers_backstore/Business/Services/ProductsService.cs
using glovers_backstore.Data;
using glovers_backstore.Data.Models;
using glovers_backstore.Services.Base;
using Microsoft.EntityFrameworkCore;
using glovers_backstore.Interfaces;

namespace glovers_backstore.Services
{
    public class ProductsService : BaseApiService, IProductsService
    {
        private readonly StoreDbContext _dbContext;

        public ProductsService(StoreDbContext dbContext)
        {
            _dbContext = dbContext;
        }

        public async Task<List<Product>> GetAllProducts()
            => await _dbContext.Products.ToListAsync();

        public async Task<Product> GetProduct(int id)
             => await _dbContext.Products.Where(p => p.Id.Equals(id)).FirstOrDefaultAsync();
        

        public async Task<List<Product>> GetLimitedProducts(int amount)
            => await _dbContext.Products.Take(amount).ToListAsync();


        public async Task<List<string>> GetAllCategories()
             => await _dbContext.Products.GroupBy(p => p.Category).Select(c => c.Key).ToListAsync();

        public async Task<List<Product>> GetAllProductsInCategory(string category)
            => await _dbContext.Products.Where(p => p.Category.Equals(category)).ToListAsync();

        public async Task<List<ProductReview>> GetAllReviewsForProduct(int productId)
            => await _dbContext.ProductReviews.Where(pr => pr.ProductId.Equals(productId)).ToListAsync();

    }
}
